File: practical04/arrayIntegral.c
#include <stdio.h>
#include <math.h>

/* Declaring functions */
float degtorad(float arg);
float tanintegral(int arg);

/* Declaring global variables */
float tanarray[13], radang[13];

float main(void) {

/* Declaring variables local to main */
    float degang[13];
    float area;
    int i, N;

/* Loop to create tan(x) from 0 to 60 degrees */
    for (i=0; i<13; i++) {
        degang[i] = i*5;
        radang[i] = degtorad(degang[i]);
        tanarray[i] = tan(radang[i]);
    }

/* Area calculation and printing */
    area = tanintegral(12);
    printf("The calculated area is %.4f\n", area);
    return 0;
}

/* Function to convert from degrees to radians */
float degtorad(float arg) {
    return( (M_PI * arg) / 180 );
}

/* Function to calculate the area based on # of points */
float tanintegral(int arg) {

/* Declaring variables local to tanintegral  */
    float calc;
    int i;
   
/* Adding heights of both sides of each trapezoid */
    for (i=0; i<arg; i++) {
        calc += tanarray[i];
        calc += tanarray[i + 1];
    }

/* Multiplying by by width of trapezoids, triangle area correction */
    calc = calc * ( (radang[arg] - radang[0]) / (2 * arg) );

    return calc;
    
}

File: practical04/arrayIntegral.f90
! Module 1 contains global variables
module mod1
    real (kind=4) :: tanarray(13), radarray(13)
end module mod1

program arrayintegral
    use mod1

! Declaring variables and parameters
    integer :: i, N
    real (kind=4) :: degarray(13), calc

! Loop to create array values
    i = 0    
    do while (i < 13)
        degarray(i+1) = i*5
        call degtorad(degarray(i+1),radarray(i+1))
        tanarray(i+1) = tan(radarray(i+1))
        i = i+1
    end do

! Calling integration function
    N = 12
    call tanintegral(N, calc)
    write(*,*) 'The calculated area is: ',calc
end program arrayintegral

! Subroutine to convert degrees to radians
subroutine degtorad(arg,arg2)
    real (kind=4), intent(in) :: arg
    real (kind=4), intent(out) :: arg2
    real (kind=4) :: pi=3.1415927
    arg2=(pi*arg)/180
end subroutine degtorad

! Subroutine to calculate area based on # of points
subroutine tanintegral(arg, arg2)
    use mod1
    integer :: i
    integer, intent(in) :: arg
    real (kind=4), intent(out) :: arg2
    
! Loop to add heights of trapezoids
    i = 1
    arg2 = 0.0
    do while (i < arg+1)
        arg2 = arg2 + tanarray(i) + tanarray(i+1)
        i = i+1
    end do

! Area correction to value
    arg2 = arg2 * ( (radarray(arg+1) - radarray(1) ) / (2*arg) )
end subroutine tanintegral
